### PR TITLE
docs: document development process

### PR DESCRIPTION
**Pull request by kuisathaverat**
This PR completes the documentation about how to build the Docker images, what Docker images are here, what it is the use of every Docker image, and how to make updates.

Closes https://github.com/elastic/observability-robots/issues/1547

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,20 +1,45 @@
...
-`mips32` is not available in `Debian 11`.
-
-## Docker Repo
-
-`docker.elastic.co/beats-dev/golang-crossbuild:[TAG]`
+# golang-crossbuild Docker images
+
+This repository contains Dockerfiles for cross building Go binaries for various platforms.
+The aim is to provide a simple way to build Go binaries for multiple platforms without having to install and configure cross compilers on your host machine.
+To do that the project provide a set of Docker images that can be used to build Go binaries for the following platforms:
```
Comment (v1v) [inline README.md:7]:
```suggestion
To do that the project provides a set of Docker images that can be used to build Go binaries for the following platforms:
```

File: README.md [hunk 2]
```diff
@@ -1,20 +1,45 @@
...
+* linux/mips64
+* linux/ppc64
+* linux/ppc64le
+* linux/s390x
+* windows/amd64
+* darwin/amd64
+* darwin/arm64
+
+The Docker images are based on Debian and the cross compilers are installed using the crossbuild-essential package.
+Each architecture have its own folder with the files needed to build the Docker image for that architecture.
```
Comment (v1v) [inline README.md:25]:
```suggestion
Each architecture has its own folder with the files needed to build the Docker image for that architecture.
```

File: README.md [hunk 3]
```diff
@@ -1,20 +1,45 @@
...
+Each architecture has its own Dockerfile to builld a Docker image for that architecture.
```
Comment (v1v) [inline README.md:26]:
```suggestion
Each architecture has its own Dockerfile to build a Docker image for that architecture.
```

File: README.md [hunk 4]
```diff
@@ -1,20 +1,45 @@
...
+Each architecture Dockerfile install the crossbuild-essential package for that architecture and the libraries needed to build our binaries.
```
Comment (v1v) [inline README.md:27]:
```suggestion
Each architecture Dockerfile installs the crossbuild-essential package for that architecture and the libraries needed to build our binaries.
```

File: README.md [hunk 5]
```diff
@@ -1,20 +1,45 @@
...
+These Dockerfiles are generated files from `Dockerfile.tmpl` template file that is in the architectures folder.
+This template is processed using a Makefile that is in the architectures folder.
+Each architecture folder has a `roofs` folder that contains the files that will be copied to the Docker image in the root folder.
+In `rootfs` we have the `compolers.yml` file that contains the list of compilers that will be installed in the Docker image.
+Each Docker image has a basic compilation test that is executed when the image is built. This test uses `rootfs/helloworld.c` file to compile a simple C program and verify the architecture of the result binary.
+The compiler used to build the binaries is LLVM.
+Some of the Docker images are build for the amd64 and arm64 architectures, this allow to run the Docker images in linux/amd64, linux/arm64, darwin/amd64, and darwin/arm64. This is done using the `.ci/scripts/buildx.sh` command.
+
+The Docker images are tagger using the following format:
```
Comment (v1v) [inline README.md:36]:
```suggestion
The Docker images are tagged using the following format:
```

File: README.md [hunk 6]
```diff
@@ -1,20 +1,45 @@
...
+
+* `docker.elastic.co/beats-dev/golang-crossbuild:<go-version>-<arch>-<debian-version>`
+
+For the latest version of the images based on the latest Debian and Go versions, the following tags are also used:
```
Comment (v1v) [inline README.md:40]:
```suggestion
For the latest version of the images based on the latest Debian and Go versions, the following tag is also used:
```

File: README.md [hunk 7]
```diff
@@ -73,24 +98,111 @@ Until Golang version 1.15
...
-docker run -it --rm \
-  -v $GOPATH/src/github.com/user/go-project:/go/src/github.com/user/go-project \
-  -w /go/src/github.com/user/go-project \
-  -e CGO_ENABLED=1 \
-  docker.elastic.co/beats-dev/golang-crossbuild:1.16.7-armhf \
-  --build-cmd "make build" \
-  -p "linux/armv7"
+There are several Makefiles in the profect across the different folders.
+The Makefile in the root folder is used to build the Docker images for the different architectures,
+it triggers the build of all dockerimages for all architectures and Devian versions supported.
```
Comment (v1v) [inline README.md:105]:
```suggestion
it triggers the build of all Docker images for all architectures and Debian versions supported.
```

File: README.md [hunk 8]
```diff
@@ -73,24 +98,111 @@ Until Golang version 1.15
...
+
+* `go/Makefile.debian7`
+* `go/Makefile.debian8`
+* `go/Makefile.debian9`
+* `go/Makefile.debian10`
+* `go/Makefile.debian11`
+
+No all architectures are supported in all Debian versions, so the Makefile for each Debian version will only build the Docker images for the architectures that are supported in that Debian version.
+
+On the Makefiles threre are some variables to define the name of the Docker image, the version of the Docker image, the Debian version, and the suffix to use in the tag of the Docker image.
```
Comment (v1v) [inline README.md:118]:
```suggestion
On the Makefiles there are some variables to define the name of the Docker image, the version of the Docker image, the Debian version, and the suffix to use in the tag of the Docker image.
```

File: README.md [hunk 9]
```diff
@@ -73,24 +98,111 @@ Until Golang version 1.15
...
+```make
+TAG := $(REPOSITORY)/$(NAME):$(VERSION)$(SUFFIX)$(TAG_EXTENSION)
+```
+
+The common variables to all the Makefiles are defined at `Makefile.common` file.
+
+To more information about the supported architextures and the correlation with comercial names,
+you can check the [Debian Supported Architectures](https://wiki.debian.org/SupportedArchitectures) page.
+
+To make and push the Docker images for all the architectures and defaulr Debian version, you can run the following command:
```
Comment (v1v) [inline README.md:145]:
```suggestion
To make and push the Docker images for all the architectures and default Debian version, you can run the following command:
```

File: README.md [hunk 10]
```diff
@@ -73,24 +98,111 @@ Until Golang version 1.15
...
+
+Finally, to build a single Docker image for `arm` architecture and Debian 10, you can run the following command:
+
+```shell
+make -C go -f Makefile.debian10 build push IMAGES=arm
+```
+
+## Multiarch Docker images
+
+Some of the Docker images are build for the amd64 and arm64 architectures, this allow to run the Docker images in linux/amd64, linux/arm64, darwin/amd64, and darwin/arm64. This is done using the `.ci/scripts/buildx.sh` command.
```
Comment (v1v) [inline README.md:165]:
```suggestion
Some of the Docker images are built for the amd64 and arm64 architectures, this allow to run the Docker images in linux/amd64, linux/arm64, darwin/amd64, and darwin/arm64. This is done using the `.ci/scripts/buildx.sh` command.
```

File: README.md [hunk 11]
```diff
@@ -73,24 +98,111 @@ Until Golang version 1.15
...
+ifeq ($(DOCKER_MULTIARCH),1)
+DOCKER_CMD := $(SELF_DIR)/../.ci/scripts/buildx.sh
+endif
+```
+
+## Docker images dependencies
+
+The Docker images depends on each other, so there is an order to build them.
+The are two Docker images that are parent of all the other images, the `fpm` and the `go/llvm-apple`.
+Anytime a new Devian version is released, the `fpm` and `go/llvm-apple` image needs to be updated to use the new Debian version.
```
Comment (v1v) [inline README.md:178]:
```suggestion
Anytime a new Debian version is released, the `fpm` and `go/llvm-apple` image needs to be updated to use the new Debian version.
```

File: README.md [hunk 12]
```diff
@@ -73,24 +98,111 @@ Until Golang version 1.15
...
+    base --> mips
+    base --> mips32
+    base --> npcap
+    base --> ppc
+    base --> s390x
+```
 
 ## Releasing images for a new Go version
 
+With every new version of `go` we made a new branch with the name of the previous version to allow continue building the Docker images for the previous version of `go`. So if we are in go `1.17` and go `1.18` is released, we create a new branch `1.17`, the we update the `master` branch to install go `1.18`. Due to the changes in the Debian packages repositories, there is no guaranties that the Docker images for the previous version of `go` will continue to work after some time.
```
Comment (v1v) [inline README.md:204]:
```suggestion
With every new version of `go` we made a new branch with the name of the previous version to allow continue building the Docker images for the previous version of `go`. So if we are in go `1.19` and go `1.20` is released, we create a new branch `1.19`, the we update the `main` branch to install go `1.20`. Due to the changes in the Debian packages repositories, there is no guaranties that the Docker images for the previous version of `go` will continue to work after some time.
```

**Review by v1v — COMMENTED** [review]
Some minor NITs and typos. Otherwise, LGTM, thanks!!

**Review by v1v — APPROVED** [review]

**Review by cachedout — APPROVED** [review]

**Comment by cachedout** [comment]
@jlind23 This is related to the transition of golang-crossbuild to Ingest. Can you please assign some reviewer(s) from Ingest to have a look at this? I would like to know if they (and you) find this documentation sufficient or if we should also schedule a call to review the process and answer questions. Thanks.

**Comment by jlind23** [comment]
@cachedout I just discussed this with @alexsapran and it makes sense for us to take this over, however our bandwidth is very limited until we hire a new Eng Prod engineer. This req is ongoing but no one has joined yet. Would be great if your team can hold the fort until further notice.

cc @amitkanfer

**Comment by cachedout** [comment]
@jlind23 That is fine. Thanks for letting me know.

**Pull request merged**